File: internal/models/adaptive_learning.go
package models

import "time"

// AdaptiveStageSettings is a concrete execution choice, never a permission to
// invent a model, prompt, precision or window. Only qualified settings may learn.
type AdaptiveStageSettings struct {
	Device           string  `json:"device"`
	Precision        string  `json:"precision"`
	BatchSize        int     `json:"batch_size"`
	Concurrency      int     `json:"concurrency"`
	WindowSeconds    float64 `json:"window_seconds"`
	OverlapSeconds   float64 `json:"overlap_seconds"`
	StitchingVersion string  `json:"stitching_version,omitempty"`
}

// Scope contains fingerprints and workload classes, not audio, transcripts,
// prompts, credentials, paths, model download URLs or exception strings.
type AdaptiveLearningScope struct {
	StageKey              string  `json:"stage_key"`
	FixedSettingsHash     string  `json:"fixed_settings_hash"`
	RuntimeFingerprint    string  `json:"runtime_fingerprint"`
	ModelFingerprint      string  `json:"model_fingerprint"`
	HardwareFingerprint   string  `json:"hardware_fingerprint"`
	WorkloadClass         string  `json:"workload_class"`
	MemoryDomain          string  `json:"memory_domain"`
	MemoryCapacityBytes   int64   `json:"memory_capacity_bytes"`
	OriginalWindowSeconds float64 `json:"original_window_seconds"`
}

type AdaptiveObservation struct {
	ID                       string                `json:"id" gorm:"primaryKey"`
	ProfileID                string                `json:"profile_id" gorm:"not null;index:idx_adaptive_observation_scope"`
	ProfileRevision          int64                 `json:"profile_revision" gorm:"not null;index:idx_adaptive_observation_scope"`
	LearningGeneration       int64                 `json:"learning_generation" gorm:"not null;index:idx_adaptive_observation_scope"`
	ScopeKey                 string                `json:"scope_key" gorm:"not null;index:idx_adaptive_observation_scope"`
	Scope                    AdaptiveLearningScope `json:"scope" gorm:"serializer:json;type:text;not null"`
	ExecutionID              string                `json:"execution_id" gorm:"not null;index"`
	StageID                  string                `json:"stage_id" gorm:"not null"`
	AttemptID                string                `json:"attempt_id" gorm:"not null;uniqueIndex"`
	RecordingID              string                `json:"recording_id" gorm:"not null"`
	OwnerGeneration          int64                 `json:"owner_generation" gorm:"not null"`
	CandidateHash            string                `json:"candidate_hash" gorm:"not null;index"`
	Settings                 AdaptiveStageSettings `json:"settings" gorm:"serializer:json;type:text;not null"`
	Outcome                  string                `json:"outcome" gorm:"not null"`
	FullStage                bool                  `json:"full_stage"`
	Cached                   bool                  `json:"cached"`
	Qualified                bool                  `json:"qualified"`
	ExternalContention       bool                  `json:"external_contention"`
	PeakMemoryBytes          *int64                `json:"peak_memory_bytes,omitempty"`
	AvailableBeforeBytes     *int64                `json:"available_before_bytes,omitempty"`
	ReserveBytes             *int64                `json:"reserve_bytes,omitempty"`
	LoadingMilliseconds      *int64                `json:"loading_milliseconds,omitempty"`
	ProcessingMilliseconds   *int64                `json:"processing_milliseconds,omitempty"`
	SourceDurationSeconds    float64               `json:"source_duration_seconds"`
	ProcessedDurationSeconds float64               `json:"processed_duration_seconds"`
	Channels                 int                   `json:"channels"`
	CreatedAt                time.Time             `json:"created_at"`
}

// Plans and their supporting observation IDs are immutable. Selection is kept
// separately so resetting learning never rewrites evidence used by an old run.
type AdaptiveLearnedPlan struct {
	ID                         string                  `json:"id" gorm:"primaryKey"`
	ProfileID                  string                  `json:"profile_id" gorm:"not null;index"`
	ProfileRevision            int64                   `json:"profile_revision" gorm:"not null;index"`
	LearningGeneration         int64                   `json:"learning_generation" gorm:"not null;index"`
	ScopeKey                   string                  `json:"scope_key" gorm:"not null;index"`
	Scope                      AdaptiveLearningScope   `json:"scope" gorm:"serializer:json;type:text;not null"`
	CandidateHash              string                  `json:"candidate_hash" gorm:"not null"`
	Settings                   AdaptiveStageSettings   `json:"settings" gorm:"serializer:json;type:text;not null"`
	ObservationIDs             []string                `json:"observation_ids" gorm:"serializer:json;type:text;not null"`
	ExhaustionEvidenceIDs      []string                `json:"exhaustion_evidence_ids,omitempty" gorm:"serializer:json;type:text"`
	FullWindowCandidateHashes  []string                `json:"full_window_candidate_hashes,omitempty" gorm:"serializer:json;type:text"`
	ExhaustionAvailableBytes   int64                   `json:"exhaustion_available_bytes,omitempty"`
	ExhaustionCapacityLimits   []AdaptiveCapacityLimit `json:"exhaustion_capacity_limits,omitempty" gorm:"serializer:json;type:text"`
	PeakMemoryBytes            int64                   `json:"peak_memory_bytes"`
	MinimumReserveBytes        int64                   `json:"minimum_reserve_bytes"`
	MeanProcessingMilliseconds int64                   `json:"mean_processing_milliseconds"`
	CreatedAt                  time.Time               `json:"created_at"`
}

// A shorter-window shortcut must still fit while every exhausted device has
// no more available memory than in its comparable full-window failure.
type AdaptiveCapacityLimit struct {
	MemoryDomain        string `json:"memory_domain"`
	MemoryCapacityBytes int64  `json:"memory_capacity_bytes"`
	MaxAvailableBytes   int64  `json:"max_available_bytes"`
}

type AdaptiveCapacityReading struct {
	MemoryDomain        string `json:"memory_domain"`
	MemoryCapacityBytes int64  `json:"memory_capacity_bytes"`
	AvailableBytes      int64  `json:"available_bytes"`
}

type AdaptivePlanSelection struct {
	ProfileID          string    `json:"profile_id" gorm:"primaryKey"`
	ProfileRevision    int64     `json:"profile_revision" gorm:"primaryKey"`
	LearningGeneration int64     `json:"learning_generation" gorm:"primaryKey"`
	ScopeKey           string    `json:"scope_key" gorm:"primaryKey"`
	PlanID             string    `json:"plan_id"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type AdaptivePlanSnapshot struct {
	PlanID             string                `json:"plan_id"`
	ScopeKey           string                `json:"scope_key"`
	ProfileRevision    int64                 `json:"profile_revision"`
	LearningGeneration int64                 `json:"learning_generation"`
	StageKey           string                `json:"stage_key"`
	CandidateHash      string                `json:"candidate_hash"`
	Settings           AdaptiveStageSettings `json:"settings"`
}

// Profile revision parameters are private profile data, with credentials
// removed before persistence. They are not copied into learning observations.
type AdaptiveProfileRevision struct {
	ID                 string         `json:"id" gorm:"primaryKey"`
	ProfileID          string         `json:"profile_id" gorm:"not null;uniqueIndex:idx_adaptive_profile_revision"`
	Revision           int64          `json:"revision" gorm:"not null;uniqueIndex:idx_adaptive_profile_revision"`
	LearningGeneration int64          `json:"learning_generation"`
	Name               string         `json:"name"`
	Description        *string        `json:"description,omitempty"`
	IsDefault          bool           `json:"is_default"`
	Parameters         WhisperXParams `json:"parameters" gorm:"serializer:json;type:text;not null"`
	Reason             string         `json:"reason"`
	SourceRevision     *int64         `json:"source_revision,omitempty"`
	SourcePlanID       *string        `json:"source_plan_id,omitempty"`
	SavedAt            time.Time      `json:"saved_at"`
}

File: web/frontend/src/features/transcription/hooks/adaptiveLearning.ts
export interface AdaptivePlanSettings {
    device: string; precision: string; batch_size: number; concurrency: number;
    window_seconds: number; overlap_seconds: number; stitching_version?: string;
}
export interface AdaptiveLearningScope {
    stage_key: string; workload_class: string; memory_domain: string; memory_capacity_bytes?: number;
    fixed_settings_hash?: string; runtime_fingerprint?: string; model_fingerprint?: string; hardware_fingerprint?: string;
}
export interface AdaptiveLearnedPlan {
    id: string; profile_revision: number; learning_generation: number; scope_key: string;
    scope: AdaptiveLearningScope; settings: AdaptivePlanSettings; observation_ids: string[];
    peak_memory_bytes?: number; minimum_reserve_bytes?: number; mean_processing_milliseconds?: number; created_at: string;
}
export interface AdaptiveObservation {
    id: string; execution_id: string; recording_id: string; attempt_id: string;
    profile_revision?: number; learning_generation?: number;
    scope_key: string; scope: AdaptiveLearningScope; settings: AdaptivePlanSettings;
    full_stage: boolean; cached: boolean; qualified: boolean; external_contention: boolean; outcome: string;
    peak_memory_bytes?: number; available_before_bytes?: number; reserve_bytes?: number;
    loading_milliseconds?: number; processing_milliseconds?: number; created_at: string;
}
export interface AdaptiveProfileRevision {
    revision: number; saved_at: string; reason: string; source_revision?: number; source_plan_id?: string;
    name: string; parameters: Record<string, unknown>;
}
export interface AdaptiveLearningResponse {
    profile_id: string; profile_revision: number; learning_generation: number; status: string;
    selected_plan_id?: string; selected_plan_ids?: string[]; plans: AdaptiveLearnedPlan[]; observations: AdaptiveObservation[];
    revisions: AdaptiveProfileRevision[];
}

export function measuredMemory(bytes?: number | null): string {
    if (bytes == null || !Number.isFinite(bytes) || bytes < 0) return "Not measured";
    return bytes >= 1024 ** 3 ? `${(bytes / 1024 ** 3).toFixed(2)} GiB` : `${(bytes / 1024 ** 2).toFixed(1)} MiB`;
}
export function measuredDuration(milliseconds?: number | null): string {
    if (milliseconds == null || !Number.isFinite(milliseconds) || milliseconds < 0) return "Not measured";
    return `${(milliseconds / 1000).toFixed(1)}s`;
}
export function adaptiveSettingsSummary(settings?: AdaptivePlanSettings): string {
    if (!settings) return "No effective settings recorded";
    return `${settings.device} · ${settings.precision} · batch ${settings.batch_size} · concurrency ${settings.concurrency} · ${settings.window_seconds > 0 ? `${settings.window_seconds}s window` : "original/full window"} · ${settings.overlap_seconds}s overlap`;
}
export function learningPlanIsCurrent(data: AdaptiveLearningResponse, plan: AdaptiveLearnedPlan): boolean {
    // Selection is current authority. Immutable evidence keeps its source
    // generation/revision when promotion or a metadata edit advances a profile.
    return Array.isArray(plan.observation_ids) && plan.observation_ids.length >= 3
        && (data.selected_plan_id === plan.id || data.selected_plan_ids?.includes(plan.id) === true);
}
export function buildLearningAction(data: AdaptiveLearningResponse, action: "reset" | "freeze" | "restore", target?: string | number): { path: string; body: Record<string, string | number> } {
    if (!Number.isSafeInteger(data.profile_revision) || data.profile_revision < 1 || !Number.isSafeInteger(data.learning_generation) || data.learning_generation < 1) throw new Error("Reload the profile before changing its learning state.");
    const prefix = `/api/v1/profiles/${encodeURIComponent(data.profile_id)}`;
    if (action === "reset") return { path: `${prefix}/reset-adaptive`, body: { expected_revision: data.profile_revision, expected_generation: data.learning_generation } };
    if (action === "freeze") {
        const plan = data.plans.find((candidate) => candidate.id === target);
        if (!plan || !learningPlanIsCurrent(data, plan)) throw new Error("Choose a measured plan from the current profile revision and learning generation.");
        return { path: `${prefix}/freeze-adaptive`, body: { expected_revision: data.profile_revision, expected_generation: data.learning_generation, plan_id: plan.id } };
    }
    if (typeof target !== "number" || !data.revisions.some((revision) => revision.revision === target) || target === data.profile_revision) throw new Error("Choose a previous saved profile revision.");
    return { path: `${prefix}/restore-revision`, body: { expected_revision: data.profile_revision, expected_generation: data.learning_generation, revision: target } };
}
